### PR TITLE
feat: add split procedure

### PR DESCRIPTION
**Pull request by ZuLiangWang**
# Which issue does this PR close?

Closes https://github.com/CeresDB/ceresmeta/issues/66

# Rationale for this change
It has been mentioned in the issue, add split procedure implementation.

# What changes are included in this PR?
* Add split procedure implementation in `split.go`.
* Implement split http interface in `api.go`.
* The capacity expansion of CeresDB node is realized through the combination of split and transferleader in `api.go`.

# Are there any user-facing changes?
None.

# How does this change test
Pass unit test.

### CODE REVIEW

File: server/cluster/topology_manager.go [hunk 1]
```diff
@@ -243,6 +246,50 @@ func (m *TopologyManagerImpl) RemoveTable(ctx context.Context, tableID storage.T
...
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if _, ok := m.shardNodesMapping[shardTables.Shard.ID]; !ok {
+		return ShardVersionUpdate{}, ErrShardNotFound.WithCausef("shard id:%d", shardTables.Shard.ID)
+	}
+
+	prevVersion := uint64(0)
+	shardView, ok := m.shardTablesMapping[shardTables.Shard.ID]
+	if ok {
```
Comment (chunshao90) [inline server/cluster/topology_manager.go:259]: If ok is false, the `prevVersion` is wrong?
Shall we return error here.

File: server/service/http/api.go [hunk 2]
```diff
@@ -178,3 +179,58 @@ func (a *API) route(writer http.ResponseWriter, req *http.Request) {
...
+		return
+	}
+
+	if err := a.procedureManager.Submit(ctx, splitProcedure); err != nil {
+		log.Error("submit split procedure", zap.Error(err))
+		a.respondError(writer, procedure.ErrSubmitProcedure, "submit split procedure")
+		return
+	}
+
+	a.respond(writer, newShardID)
```
Comment (chunshao90) [inline server/service/http/api.go:235]: Please return msg with json format.

File: server/service/http/api.go [hunk 3]
```diff
@@ -178,3 +179,58 @@ func (a *API) route(writer http.ResponseWriter, req *http.Request) {
...
+
+func (a *API) split(writer http.ResponseWriter, req *http.Request) {
+	var splitRequest SplitRequest
+	err := json.NewDecoder(req.Body).Decode(&splitRequest)
+	if err != nil {
+		log.Error("decode request body failed", zap.Error(err))
+		a.respondError(writer, ErrParseRequest, nil)
+		return
+	}
+	ctx := context.Background()
```
Comment (chunshao90) [inline server/service/http/api.go:199]: Please define ctx with timeout.

File: server/service/http/api.go [hunk 4]
```diff
@@ -178,3 +179,58 @@ func (a *API) route(writer http.ResponseWriter, req *http.Request) {
...
+
+	c, err := a.clusterManager.GetCluster(ctx, splitRequest.ClusterName)
+	if err != nil {
+		log.Error("cluster not found", zap.String("clusterName", splitRequest.ClusterName), zap.Error(err))
+		a.respondError(writer, cluster.ErrClusterNotFound, "cluster not found")
```
Comment (chunshao90) [inline server/service/http/api.go:204]:
```suggestion
		a.respondError(writer, cluster.ErrClusterNotFound, fmt.Sprintf("cluster not found, cluster:%s",splitRequest.ClusterName))
```

File: server/service/http/api.go [hunk 5]
```diff
@@ -178,3 +179,58 @@ func (a *API) route(writer http.ResponseWriter, req *http.Request) {
...
+		return
+	}
+
+	newShardID, err := c.AllocShardID(ctx)
+	if err != nil {
+		log.Error("alloc shard id failed")
```
Comment (chunshao90) [inline server/service/http/api.go:210]:
```suggestion
		log.Error("alloc shard id failed",zap.Error(err))
```

File: server/coordinator/procedure/split.go [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+	tableNames     []string
+	targetNodeName string
+	schemaName     string
+
+	// Protect the state.
+	lock  sync.RWMutex
+	state State
+}
+
+func NewSplitProcedure(id uint64, dispatch eventdispatch.Dispatch, storage Storage, c *cluster.Cluster, schemaName string, shardID storage.ShardID, newShardID storage.ShardID, tableNames []string, targetNodeName string) *SplitProcedure {
```
Comment (chunshao90) [inline server/coordinator/procedure/split.go:77]: The parameter is too long. Define a struct to hold these parameters.

File: server/coordinator/procedure/split.go [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+
+func splitOpenShardCallback(event *fsm.Event) {
+	request, err := getRequestFromEvent[splitCallbackRequest](event)
+	if err != nil {
+		cancelEventWithLog(event, err, "get request from event")
+		return
+	}
+	ctx := request.ctx
+
+	// Send open new shard request to CSE.
```
Comment (chunshao90) [inline server/coordinator/procedure/split.go:309]:
```suggestion
	// Send open new shard request to Ceresdb.
```

File: server/coordinator/procedure/split.go [hunk 8]
```diff
@@ -0,0 +1,470 @@
...
+		})
+	}
+
+	if _, err := request.cluster.UpdateShardTables(ctx, cluster.ShardTables{
+		Shard:  newShardInfo,
+		Tables: tables,
+	}); err != nil {
+		cancelEventWithLog(event, err, "update shard tables")
+		return
+	}
```
Comment (chunshao90) [inline server/coordinator/procedure/split.go:413]: Add some log info at key points.

File: server/coordinator/procedure/util.go [hunk 9]
```diff
@@ -32,3 +32,24 @@ func getRequestFromEvent[T any](event *fsm.Event) (T, error) {
...
 }
+
+func IsContains(slice []string, target string) bool {
+	for _, a := range slice {
+		if a == target {
+			return true
+		}
+	}
+	return false
+}
```
Comment (chunshao90) [inline server/coordinator/procedure/util.go:43]: Refer to https://pkg.go.dev/golang.org/x/exp/slices#IndexFunc.

File: server/coordinator/procedure/util.go [hunk 10]
```diff
@@ -32,3 +32,24 @@ func getRequestFromEvent[T any](event *fsm.Event) (T, error) {
...
+
+func IsSubSlice(subSlice []string, slice []string) bool {
```
Comment (chunshao90) [inline server/coordinator/procedure/util.go:45]: Use the hash map to determine whether the item is in the slice?

File: server/cluster/topology_manager.go (hunk 1 above)
Comment (ZuLiangWang) [reply server/cluster/topology_manager.go:259]: `ShardView` is not exists when it has not been created, this is a normal condition, and error cannot be returned.

**Review by jiacai2050 — APPROVED** [review]

**Pull request merged**